File: client/src/types/account.ts
import { ICategory } from './category';

export interface Account {
  id: string;
  syncID: string;
  name: string;
  institutionID: string;
  type: string;
  subtype: string;
  currentBalance: number;
  balanceDate: Date;
  hideTransactions: boolean;
  hideAccount: boolean;
  deleted: Date;
  index: number;
  userID: string;
}

export interface AccountEditRequest {
  id: string;
  name: string;
  type: string;
  subtype: string;
  hideTransactions: boolean;
  hideAccount: boolean;
}

export interface AccountIndexRequest {
  id: string;
  index: number;
}

export interface NewAccount extends Partial<Account> {}

export const accountCategories: ICategory[] = [
  {
    value: 'Checking',
    parent: '',
  },
  {
    value: 'Savings',
    parent: '',
  },
  {
    value: 'Money Market',
    parent: 'Savings',
  },
  {
    value: 'Certificate of Deposit',
    parent: 'Savings',
  },
  {
    value: 'Loan',
    parent: '',
  },
  {
    value: 'Auto',
    parent: 'Loan',
  },
  {
    value: 'Student',
    parent: 'Loan',
  },
  {
    value: 'Personal',
    parent: 'Loan',
  },
  {
    value: 'Home Equity',
    parent: 'Loan',
  },
  {
    value: 'Credit Card',
    parent: '',
  },
  {
    value: 'Investment',
    parent: '',
  },
  {
    value: '401k',
    parent: 'Investment',
  },
  {
    value: 'Roth IRA',
    parent: 'Investment',
  },
  {
    value: 'Rollover IRA',
    parent: 'Investment',
  },
  {
    value: 'ESPP',
    parent: 'Investment',
  },
  {
    value: 'Trust',
    parent: 'Investment',
  },
  {
    value: 'Taxable',
    parent: 'Investment',
  },
  {
    value: 'Mortgage',
    parent: '',
  },
  {
    value: 'Cash',
    parent: '',
  },
  {
    value: 'Other',
    parent: '',
  },
];

File: server/Models/Account.cs
using BudgetBoard.Database.Models;

namespace BudgetBoard.Models;

public class AccountEditRequest
{
    public Guid ID { get; set; }
    public required string Name { get; set; }
    public string Type { get; set; } = "";
    public string Subtype { get; set; } = "";
    public bool HideTransactions { get; set; } = false;
    public bool HideAccount { get; set; } = false;
}

public class AccountIndexRequest
{
    public Guid ID { get; set; }
    public int Index { get; set; }
}

public class AccountResponse
{
    public Guid ID { get; set; }
    public string? SyncID { get; set; }
    public string Name { get; set; } = string.Empty;
    public Guid? InstitutionID { get; set; }
    public string Type { get; set; } = string.Empty;
    public string Subtype { get; set; } = string.Empty;
    public float CurrentBalance { get; set; } = 0.0f;
    public DateTime? BalanceDate { get; set; } = null;
    public bool HideTransactions { get; set; } = false;
    public bool HideAccount { get; set; } = false;
    public DateTime? Deleted { get; set; } = null;
    public int Index { get; set; }
    public Guid UserID { get; set; }

    public AccountResponse(Account account)
    {
        ID = account.ID;
        Name = account.Name;
        InstitutionID = account.InstitutionID;
        Type = account.Type;
        Subtype = account.Subtype;
        CurrentBalance = account.CurrentBalance;
        BalanceDate = account.BalanceDate;
        HideTransactions = account.HideTransactions;
        HideAccount = account.HideAccount;
        Deleted = account.Deleted;
        Index = account.Index;
        UserID = account.UserID;
    }
}
